src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -859,6 +911,99 @@ private static void OnPingComplete(object sender, PingCompletedEventArgs e)
             ((TestConnectionCommand)e.UserState)._pingComplete.Set();
         }
 
+        /// <summary>
+        /// The class contains information about the TCP connection test
+        /// </summary>
+        public class TcpTestStatus
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="TcpTestStatus"/> class.
+            /// </summary>
+            /// <param name="testNum">The number of this test.</param>
+            /// <param name="source">The source machine name or IP of the test.</param>
+            /// <param name="destination">The destination machine name or IP of the test.</param>
+            /// <param name="destinationAddress">The resolved IP from the destination</param>
+            /// <param name="port">The port used for the connection.</param>
+            /// <param name="latency">The latency of the test.</param>
+            /// <param name="result">The result of the test.</param>
+            internal TcpTestStatus(int testNum, string source, string destination, string destinationAddress, int port, long latency, TcpConnectionTestResult result)
+            {
+                TestNum = testNum;
+                Source = source;
+                Destination = destination;
+                DestinationAddress = destinationAddress;
+                Port = port;
+                Latency = latency;
+                Result = result;
+            }
+
+            /// <summary>
+            /// Gets and sets the count of the test.
+            /// </summary>
+            public int TestNum { get; set; }
+
+            /// <summary>
+            /// Gets the source from which the test was sent.
+            /// </summary>
+            public string Source { get; }
+
+            /// <summary>
+            /// Gets the destination name.
+            /// </summary>
+            public string Destination { get; }
+
+            /// <summary>
+            /// Gets the resolved address for the destination.
+            /// </summary>
+            public string DestinationAddress { get; }
+
+            /// <summary>
+            /// Gets the port used for the test.
+            /// </summary>
+            public int Port { get; }
+
+            /// <summary>
+            /// Gets or sets the latancy of the connection
+            /// </summary>
+            public long Latency { get; set; }
+
+            /// <summary>
+            /// Gets or sets the result of the test
+            /// </summary>
+            public TcpConnectionTestResult Result { get; set; }
```

Comment:
I'd consider using `IPStatus` type.

Reply:
I've looked into this a bit more! A few of the statuses line up really well, but a few are missing. Here's the mapping I'm looking at: 

| TcpConnectionTestResult | IPStatus | IPStatusDescription | 
| --- | --- | --- | 
| New | ? |  | 
| Success | Success | The ICMP echo request succeeded; an ICMP echo reply was received. | 
| Failed | Unknown | The ICMP echo request failed for an unknown reason. | 
| Timeout | TimedOut | The ICMP echo Reply was not received within the allotted time. | 
| Cancelled | ? |  | 

On concern I have about this, is the output may be a bit misleading. The errors specified in the `IPStatus` type are fairly icmp specific, and don't translate 1:1 with tcp. 

What do you reckon in this case? Thanks!!

Reply:
Do we output an object when it is `Cancelled`? From what I remember of that part of the code, the action we take on cancelling the test is simply to return nothing, isn't it? If so, we don't really need to use a `Cancelled` status.

We can use `IPStatus.Unknown` for the `New`; it's basically the "unset" value. 🙂

Reply:
That's a good point, thinking about it a bit more the `Cancelled` option is a bit redundant. It gets set if the task is cancelled, but nothing gets returned anyways 🤦‍♂️. I like the idea of using `IPStatus.Unknown` as the default "unset" value! 

I reckon we could map it out like this: 

| TcpConnectionTestResult | IPStatus |
| --- | --- | 
| New | Unknown |
| Success | Success | 
| Failed | Unknown | 
| Timeout | TimedOut |

My only concern is a failed test may be a bit misleading as "Unknown", as it's fairly ambiguous. 

What do you think about this?

Reply:
Yeah we should differentiate it if we can... What are the expected failure cases apart from a timeout? Connection refused? Something else?

We could possibly utilise the `DestinationUnreachable` mode, but I'm not seeing many other options. You're right, this enum is pretty specific to ICMP. Perhaps there's a status enum used for socket connections in .NET Core? Let me see...

Have a look at `System.Net.Sockets.SocketError` -- those are a bit more generalized and have all the states we need at the moment, I think. 🙂

Reply:
We can use `SocketError.InProgress` for the "new" state?

Reply:
@vexx32 I just scoped out `SocketError`, this looks like a much more organic fit! I've been playing around with this a bit. How do you guys feel about the idea of returning the actual error code when we run into an exception, instead of something generic? Here's a bit of code to better explain: 

```csharp
catch (AggregateException ae)
{
    ae.Handle((ex) =>
    {
        if (ex is SocketException)
        {
            SocketException? socketException = ex as SocketException;
            testResult.Result = socketException!.SocketErrorCode;
            return true;
        }
        else
        {
            testResult.Result = SocketError.SocketError;
            return false;
        }
    });
}
```

This would provide a bit more verbosity when running into any issues. Here's an example of this test running against a webserver as it reboots: 

```
   6 Jacks-MBP        jacktest.com     35.233.210.37          443      14 Success
   7 Jacks-MBP        jacktest.com     35.233.210.37          443      20 Success
   8 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
   9 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
  10 Jacks-MBP        jacktest.com     35.233.210.37          443       0 TimedOut
  11 Jacks-MBP        jacktest.com     35.233.210.37          443       0 TimedOut
  12 Jacks-MBP        jacktest.com     35.233.210.37          443       0 TimedOut
  13 Jacks-MBP        jacktest.com     35.233.210.37          443       0 TimedOut
  14 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
  15 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
  16 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
  17 Jacks-MBP        jacktest.com     35.233.210.37          443       0 ConnectionRefused
  18 Jacks-MBP        jacktest.com     35.233.210.37          443      21 Success
  19 Jacks-MBP        jacktest.com     35.233.210.37          443      21 Success
```

How do you guys feel about this approach?

Reply:
I think it is enough for us to have:
```c#
 public bool Connected { get; set; } // Result = TCPClient.Connected;
 public SocketError SocketError  { get; set; } // Optional
```

Reply:
Ah I see what you mean! 

In this case we would only set SocketError in the case that we actually run into an error? What would the default value of SocketError be? 

I really like having the two separated, it should make integration further down the pipeline more straightforward also! 😁

Reply:
Default SocketError value is 0 that is Success.

Reply:
"Result" is general name. I suggest to use more specific name like "Connected" with bool type.

Reply:
In the middle of moving, haven't had much time to dig into this for a few days 😓

@iSazonov I've played around with the idea of having a boolean value for the connection, and a more detailed value with error states. Here's an example output of what I have set up so far:

```
> Test-Connection jacktest.com -tcpport 443 -count 100

Target: jacktest.com

Id Source           Address                      Port Latency Connected  Status
                                                        (ms)
-- ------           -------                      ---- ------- ---------  ------
 9 Jacks-MBP        35.233.210.37                 443      21 True       Success
10 Jacks-MBP        35.233.210.37                 443      14 True       Success
11 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
12 Jacks-MBP        35.233.210.37                 443       0 False      TimedOut
13 Jacks-MBP        35.233.210.37                 443       0 False      TimedOut
14 Jacks-MBP        35.233.210.37                 443       0 False      TimedOut
15 Jacks-MBP        35.233.210.37                 443       0 False      TimedOut
16 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
17 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
18 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
19 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
20 Jacks-MBP        35.233.210.37                 443       0 False      ConnectionRefused
21 Jacks-MBP        35.233.210.37                 443      43 True       Success
22 Jacks-MBP        35.233.210.37                 443      23 True       Success
```
I've also removed `Target` as a column. As it's displayed initially, I thought it was a bit redundant. 

Please let me know if this is what you're thinking, and if you have any feedback!

Reply:
It looks a lot better. Thanks!